Cache gradients with the destination colorspace

smfr opened the pull request:
#### 47195378027dfc79782682ef15e877bb87cb96b9
<pre>
Cache gradients with the destination colorspace
<a href="https://bugs.webkit.org/show_bug.cgi?id=292808">https://bugs.webkit.org/show_bug.cgi?id=292808</a>
<a href="https://rdar.apple.com/151043882">rdar://151043882</a>

Reviewed by Sam Weinig.

When a gradient is painted by the underlying frameworks, its colors (usually sRGB) are converted
to the display colorspace, which has some cost. Gradients are usually fairly long-lived and painted
many times, so we can avoid that repeated colorspace conversion by having the gradient store colors
in the destination colorspace.

The test fails on iOS because of <a href="https://rdar.apple.com/151188818">rdar://151188818</a>.

* LayoutTests/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/fast/canvas/gradient-into-p3-canvas-expected.txt: Added.
* LayoutTests/fast/canvas/gradient-into-p3-canvas.html: Added. Tests the change in `platformConvertColorComponents()`.
* Source/WebCore/platform/graphics/Gradient.h: Pass an optional DestinationColorSpace to paint().
* Source/WebCore/platform/graphics/cg/ColorCG.cpp:
(WebCore::platformConvertColorComponents): We need to copy over the alpha, which is the fourth component.
* Source/WebCore/platform/graphics/cg/GradientCG.cpp:
(WebCore::Gradient::paint): The cached platform gradient can be reused if its colorspace matches the
destination colorspace.
* Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
(WebCore::GradientRendererCG::GradientRendererCG):
(WebCore::GradientRendererCG::colorSpace const):
(WebCore::GradientRendererCG::pickStrategy const):
(WebCore::GradientRendererCG::makeGradient const): Convert the colors to the destination colorspace.
* Source/WebCore/platform/graphics/cg/GradientRendererCG.h:

Canonical link: <a href="https://commits.webkit.org/294844@main">https://commits.webkit.org/294844@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dd5a97b44ab61667449b3629739661411e1f3ddc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103229 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22905 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13225 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108398 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53872 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105268 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23241 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31511 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78461 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35390 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106235 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18018 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93131 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58794 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17860 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11174 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53227 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87667 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11236 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110775 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30367 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22463 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87457 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30732 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89330 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87090 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22168 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31947 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9662 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24690 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30296 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35615 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30102 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33429 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31664 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/platform/graphics/cg/ColorCG.cpp:
```diff
@@ -227,8 +227,8 @@ ColorComponents<float, 4> platformConvertColorComponents(ColorSpace inputColorSp
     ASSERT_UNUSED(result, result);
-    // FIXME: CGColorTransformConvertColorComponents doesn't copy over any alpha component.
-    return { static_cast<float>(destinationComponents[0]), static_cast<float>(destinationComponents[1]), static_cast<float>(destinationComponents[2]), static_cast<float>(destinationComponents[3]) };
+    // CGColorTransformConvertColorComponents doesn't copy over any alpha component.
```
Did this cause any changes in behavior that should be tested?

smfr replied:
I did need this for my conversion to work. However I can't directly test the gradient changes because we set the screen to sRGB for testing, but maybe I can test a gradient drawing into a P3 canvas or something.

shallawa replied:
I think this should happen only
```
if (CGColorSpaceGetNumberOfComponents(outputColorSpace.platformColorSpace()) < 4)
```

weinig commented on Source/WebCore/platform/graphics/cg/GradientCG.cpp:
```diff
@@ -48,13 +48,19 @@ void Gradient::fill(GraphicsContext& context, const FloatRect& rect)
 {
-    if (!m_platformRenderer)
-        m_platformRenderer = GradientRendererCG { m_colorInterpolationMethod, m_stops.sorted() };
+    auto ensurePlatformRenderer = [&]() {
```
```suggestion
    auto ensurePlatformRenderer = [&] {
```

You don't need the `()` if there are no parameters.

weinig commented on Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
```diff
@@ -485,11 +493,20 @@ GradientRendererCG::Strategy GradientRendererCG::makeGradient(ColorInterpolation
             for (const auto& stop : stops) {
-                auto [r, g, b, a] = stop.color.toColorTypeLossy<SRGBA<float>>().resolved();
-                colorComponents.appendList({ r, g, b, a });
+                if (destinationColorSpace) {
```
Why is this only happening in the all bounded sRGB case?

smfr replied:
Because that happens to be the case I'm targeting for optimization.

weinig commented on Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
```diff
@@ -485,11 +493,20 @@ GradientRendererCG::Strategy GradientRendererCG::makeGradient(ColorInterpolation
+
+            if (destinationColorSpace)
+                return destinationColorSpace->platformColorSpace();
+
```
How does this all work with the OutputSpaceColorType used in GradientRendererCG::makeShading? I would have expected that needed to change as well if we weren't always going to use sRGB or extendedSRGB anymore.

smfr replied:
Using the destination colorspace is optional; it's just an optimization, so the less common shading path doesn't need to do it.

shallawa commented on Source/WebCore/platform/graphics/cg/GradientCG.cpp:
```diff
@@ -48,13 +48,19 @@ void Gradient::fill(GraphicsContext& context, const FloatRect& rect)
+        m_platformRenderer = GradientRendererCG { m_colorInterpolationMethod, m_stops.sorted(), colorSpace };
+    };
+
+    ensurePlatformRenderer();
```
Shouldn't we just replace 
```
if (!m_platformRenderer)
```
with
```
if (!m_platformRenderer || m_platformRenderer->colorSpace() != colorSpace)
```

smfr replied:
I prefer the self-contained `ensure`

shallawa commented on Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
```diff
@@ -33,11 +33,19 @@
+std::optional<DestinationColorSpace> GradientRendererCG::colorSpace() const
+{
+    if (std::holds_alternative<Gradient>(m_strategy))
+        return std::get<Gradient>(m_strategy).colorSpace;
```
Can't not we use `m_strategy.get_if<Gradient>`? In this case we do not check and get in two separate statements.

weinig reviewed (approved):
I think some additional comments in GradientRendererCG explaining that only the bounded sRGB / using sRGB interpolation case is being handled for now would be good. 

OKLab interpolation is likely to be more and more common going forward, and there is talk of making it the default for all CSS gradients, so noting the optimization doesn't work for that case may become important.

smfr committed:
Cache gradients with the destination colorspace
https://bugs.webkit.org/show_bug.cgi?id=292808
rdar://151043882

Reviewed by Sam Weinig.

When a gradient is painted by the underlying frameworks, its colors (usually sRGB) are converted
to the display colorspace, which has some cost. Gradients are usually fairly long-lived and painted
many times, so we can avoid that repeated colorspace conversion by having the gradient store colors
in the destination colorspace.

The test fails on iOS because of rdar://151188818.

* LayoutTests/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/fast/canvas/gradient-into-p3-canvas-expected.txt: Added.
* LayoutTests/fast/canvas/gradient-into-p3-canvas.html: Added. Tests the change in `platformConvertColorComponents()`.
* Source/WebCore/platform/graphics/Gradient.h: Pass an optional DestinationColorSpace to paint().
* Source/WebCore/platform/graphics/cg/ColorCG.cpp:
(WebCore::platformConvertColorComponents): We need to copy over the alpha, which is the fourth component.
* Source/WebCore/platform/graphics/cg/GradientCG.cpp:
(WebCore::Gradient::paint): The cached platform gradient can be reused if its colorspace matches the
destination colorspace.
* Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
(WebCore::GradientRendererCG::GradientRendererCG):
(WebCore::GradientRendererCG::colorSpace const):
(WebCore::GradientRendererCG::pickStrategy const):
(WebCore::GradientRendererCG::makeGradient const): Convert the colors to the destination colorspace.
* Source/WebCore/platform/graphics/cg/GradientRendererCG.h:

Canonical link: https://commits.webkit.org/294844@main
